packages/flutter/lib/src/painting/inline_linked_text.dart

```diff
@@ -0,0 +1,557 @@
+        return true;
+      }
+
+      // Remove overlapping ranges.
+      final bool overlaps = textLinkerMatch.textRange.start < lastEnd;
+      if (!overlaps) {
+        lastEnd = textLinkerMatch.textRange.end;
+      }
+      return overlaps;
+    });
+
+    return nextTextLinkerMatches;
+  }
+
+  /// Apply the given [TextLinker]s to the given [InlineSpan]s and return the
+  /// new resulting spans and any created [TapGestureRecognizer]s.
+  ///
+  /// {@macro flutter.painting.LinkBuilder.recognizer}
+  static (Iterable<InlineSpan>, Iterable<TapGestureRecognizer>) linkSpans(Iterable<InlineSpan> spans, Iterable<TextLinker> textLinkers) {
+    // Flatten the spans and find all ranges in the flat String. This must be done
+    // cumulatively, and not during a traversal, because matches may occur across
+    // span boundaries.
+    final List<({InlineSpan span, int length})> spansWithLength =
+        <({InlineSpan span, int length})>[];
+    String spansText = '';
+    for (final InlineSpan span in spans) {
+      final String string = span.toPlainText();
+      spansText += string;
+      spansWithLength.add((
+        span: span,
+        length: string.length,
+      ));
+    }
+    final Iterable<_TextLinkerMatch> textLinkerMatches =
+        _cleanTextLinkerMatches(
+          _TextLinkerMatch.fromTextLinkers(textLinkers, spansText),
+        );
+
+    final (Iterable<InlineSpan> output, Iterable<_TextLinkerMatch> _, Iterable<TapGestureRecognizer> recognizers) =
+        _linkSpansRecurse(spansWithLength, textLinkerMatches, 0);
+    return (output, recognizers);
+  }
+
+  static (Iterable<InlineSpan>, Iterable<_TextLinkerMatch>, Iterable<TapGestureRecognizer>) _linkSpansRecurse(Iterable<({InlineSpan span, int length})> spansWithLength, Iterable<_TextLinkerMatch> textLinkerMatches, int index) {
+    final List<InlineSpan> output = <InlineSpan>[];
+    Iterable<_TextLinkerMatch> nextTextLinkerMatches = textLinkerMatches;
+    final List<TapGestureRecognizer> recognizers = <TapGestureRecognizer>[];
+    int nextIndex = index;
+    for (final (span: InlineSpan span, length: int length) in spansWithLength) {
+      final (InlineSpan childSpan, Iterable<_TextLinkerMatch> childTextLinkerMatches, Iterable<TapGestureRecognizer> childRecognizers) = _linkSpanRecurse(
+        span,
+        nextTextLinkerMatches,
+        nextIndex,
+      );
+      output.add(childSpan);
+      nextTextLinkerMatches = childTextLinkerMatches;
+      recognizers.addAll(childRecognizers);
+      nextIndex += length;
+    }
+
+    return (output, nextTextLinkerMatches, recognizers);
+  }
+
+  // index is the index of the start of `span` in the overall flattened tree
+  // string.
+  //
+  // The TapGestureRecognizers must be disposed by an owning widget.
+  static (InlineSpan, Iterable<_TextLinkerMatch>, Iterable<TapGestureRecognizer>) _linkSpanRecurse(InlineSpan span, Iterable<_TextLinkerMatch> textLinkerMatches, int index) {
```

Comment:
The return type and the implementation is a bit hard to read I wonder if this can be simplified.

Reply:
1. All of the caching logic has been moved into _TextCache.
  2. All of this recursive code has been moved into its own class (_LinkedSpans).

Let me know what you think.  Maybe I could still clean up all of the big records.